src/Network/Wai/Handler/Warp/Systemd.hs

```diff
@@ -230,7 +230,8 @@ runSystemdWarp saSettings settings app = do
                                void Systemd.notifyStopping
                                closeListenSocket
                                _onBeginShutdown saSettings
-                         in void $ Signals.installHandler Signals.sigTERM handler Nothing
+                         in forM_ [Signals.sigINT, Signals.sigTERM] $ \signal ->
+                              Signals.installHandler signal handler Nothing
```

Comment:
Looks like this function does both too much and too little.
I think it should return the handler somehow, or there should be a lower level function that exposes this.
Otoh it throws away the old handler. It should probably still invoke that, using `mfix` to bring it into scope.
Other than that, this is already an improvement.

Reply:
One thing I've noticed is that `onBeginShutdown` will likely not be run at all in most apps. If `runSystemdWarp` is the last call on the main thread, it'll return once `closeListenSocket` is called and the process will exit, cutting off whatever is happening in `onBeginShutdown`. It's best to instead run any cleanup functions right after `runSystemdWarp`.

Signal handling is tricky business. If you're doing cleanup after the server shuts down, then you don't really want the previous handler (which likely calls `UserInterrupt`) to run. Then again,  `CatchOnce` might be the better option here to close the socket once and then allow further sigINTs to take down the process.

Reply:
Or it could `CatchOnce` and then reinstall the handler.

I guess there isn't a one size fits all solution here, especially if the warp server isn't the only relevant service in the process.

Probably needs splitting into lower level functions as said, and keep the big function around for making the simple case easy.